misc/jobserver_test.py

```diff
@@ -0,0 +1,308 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from textwrap import dedent
+import os
+import platform
+import subprocess
+import tempfile
+import typing as T
+import sys
+import unittest
+
+_SCRIPT_DIR = os.path.realpath(os.path.dirname(__file__))
+_JOBSERVER_POOL_SCRIPT = os.path.join(_SCRIPT_DIR, "jobserver_pool.py")
+_JOBSERVER_TEST_HELPER_SCRIPT = os.path.join(_SCRIPT_DIR, "jobserver_test_helper.py")
+
+_PLATFORM_IS_WINDOWS = platform.system() == "Windows"
+
+default_env = dict(os.environ)
+default_env.pop("NINJA_STATUS", None)
+default_env.pop("MAKEFLAGS", None)
+default_env["TERM"] = "dumb"
+NINJA_PATH = os.path.abspath("./ninja")
+
+
+class BuildDir:
+    def __init__(self, build_ninja: str):
+        self.build_ninja = dedent(build_ninja)
+        self.d = None
+
+    def __enter__(self):
+        self.d = tempfile.TemporaryDirectory()
+        with open(os.path.join(self.d.name, "build.ninja"), "w") as f:
+            f.write(self.build_ninja)
+            f.flush()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.d.cleanup()
+
+    @property
+    def path(self) -> str:
+        return self.d.name
+
+    def run(
+        self,
+        cmd_flags: T.Sequence[str] = [],
+        env: T.Dict[str, str] = default_env,
+    ) -> None:
+        """Run a command, raise exception on error. Do not capture outputs."""
+        ret = subprocess.run(cmd_flags, env=env)
+        ret.check_returncode()
+
+    def ninja_run(
+        self,
+        ninja_args: T.Sequence[str],
+        prefix_args: T.Sequence[str] = [],
+        extra_env: T.Dict[str, str] = {},
+    ) -> "subprocess.CompletedProcess[str]":
+        ret = self.ninja_spawn(
+            ninja_args,
+            prefix_args=prefix_args,
+            extra_env=extra_env,
+            capture_output=False,
+        )
+        ret.check_returncode()
+        return
+
+    def ninja_clean(self):
+        self.ninja_run(["-t", "clean"])
+
+    def ninja_spawn(
+        self,
+        ninja_args: T.Sequence[str],
+        prefix_args: T.Sequence[str] = [],
+        extra_env: T.Dict[str, str] = {},
+        capture_output: bool = True,
+    ) -> "subprocess.CompletedProcess[str]":
+        """Run Ninja command and capture outputs."""
+        env = None
+        if extra_env:
+            env = default_env.copy()
+            env.update(extra_env)
+        return subprocess.run(
+            prefix_args + [NINJA_PATH, "-C", self.path] + ninja_args,
+            text=True,
+            capture_output=capture_output,
+            env=env,
```

Comment:
Fantastic tip. This is so elegant. Thank you.

Reply:
I suppose that you are in good company for not knowing about it then -- the official word is that it looks ugly and confusing and no one should be expected to assume it would work :)

https://peps.python.org/pep-0584/#motivation

Since python 3.9 you can do `env = default_env | extra_env`, which is completely intuitive, especially the part where everyone intuitively knows it will work on all supported versions of python. Oh right! :D

Still, it is elegant to have a union operator when your environment supports it...

Reply:
Wow, thanks for the history and the PEP link.
It turns out that our CI bots are running some builds on Ubuntu 20.00 (which uses Python 3.8) and on Rocky Linux 8 (which uses Python 3.6!), so I propose we just stick to dictionary unpacking for now :-/